src/plugins/ssh/index.ts

```diff
@@ -57,19 +57,28 @@ const accessPropagationGuard = (
 ) => {
   let isEphemeralAccessDeniedException = false;
   let isLoginException = false;
+  let isValidPreTestError = false;
 
   child.stderr.on("data", (chunk) => {
     const chunkString: string = chunk.toString("utf-8");
     parseAndPrintSshOutputToStderr(chunkString, options);
 
-    const match = provider.unprovisionedAccessPatterns.find((message) =>
-      chunkString.match(message.pattern)
+    const matchUnprovisionedPattern = provider.unprovisionedAccessPatterns.find(
+      (message) => chunkString.match(message.pattern)
     );
 
-    if (match) {
+    const matchPreTestPattern = provider.validPreTestAccessPatterns?.find(
+      (message) => chunkString.match(message.pattern)
+    );
```

Comment:
I wonder if `accessPropagationGuard` has to "know" about pre-testing.

There could be a new arg `validAccessPatterns` passed to `accessPropagationGuard` to make this more general. The `isAccessPropagated()` returned function can take into. account the pre-test pattern so the caller in `src/plugins/ssh/index.ts` doesn't need to know about pre-testing either.

(Also, `provider` arg could be ultimately removed by just taking patterns.)

```
const accessPropagationGuard = (
  child: ChildProcessByStdio<null, null, Readable>,
  invalidAccessPatterns: readonly AccessPattern[],
  validAccessPatterns: readonly AccessPattern[] | undefined,
  loginRequiredPattern: RegExp | undefined,
  options: SpawnSshNodeOptions
) => {
  let isEphemeralAccessDeniedException = false;
  let isLoginException = false;
  let isValidError = false;

  child.stderr.on("data", (chunk) => {
    const chunkString: string = chunk.toString("utf-8");
    parseAndPrintSshOutputToStderr(chunkString, options);

    const matchUnprovisionedPattern = invalidAccessPatterns.find((message) =>
      chunkString.match(message.pattern)
    );

    const matchPreTestPattern = validAccessPatterns?.find((message) =>
      chunkString.match(message.pattern)
    );

    if (matchUnprovisionedPattern) {
      isEphemeralAccessDeniedException = true;
    }

    if (matchPreTestPattern && !matchUnprovisionedPattern) {
      isValidError = true;
    }

    if (loginRequiredPattern) {
      const loginMatch = chunkString.match(loginRequiredPattern);
      isLoginException = isLoginException || !!loginMatch; // once true, always true
    }

    if (isLoginException) {
      isEphemeralAccessDeniedException = false; // always overwrite to false so we don't retry the access
    }
  });

  return {
    isAccessPropagated: () =>
      !isEphemeralAccessDeniedException &&
      (!validAccessPatterns || isValidError),
    isLoginException: () => isLoginException,
  };
};
```
`src/plugins/ssh/index.ts`:
```
...
    const { isAccessPropagated, isLoginException } = accessPropagationGuard(
      child,
      provider.unprovisionedAccessPatterns,
      provider.validPreTestAccessPatterns,
      provider.loginRequiredPattern,
      options
    );
...
      options.abortController?.abort(code);
      if (options.isAccessPropagationPreTest && isAccessPropagated()) {
        // override the exit code to 0 if the expected error was found, this means access is ready.
        resolve(0);
        return;
      }
      print2(`SSH session terminated`);
      resolve(code);
```

Reply:
i like it! 👍 fixed.